deltacat/compute/stats/utils/io.py

```diff
@@ -0,0 +1,114 @@
+        if stats_completion_info_s3:
+            cache_hit_stats_completion_info.append(stats_completion_info_s3)
+        else:
+            cache_miss_delta_locators.append(delta_locator)
+
+    return DeltaStatsCacheResult.of(cache_hit_stats_completion_info, cache_miss_delta_locators)
+
+
+@ray.remote
+def cache_delta_stats(stat_results_s3_bucket: str,
+                      stats_completion_info: StatsCompletionInfo):
+    scf.write_stats_completion_file(stat_results_s3_bucket, stats_completion_info)
+
+
+@ray.remote
+def get_delta_stats(delta_locator: DeltaLocator,
+                    stat_types_to_collect: Set[StatsType],
+                    deltacat_storage=unimplemented_deltacat_storage) -> StatsCompletionInfo:
+
+    delta_stats = {}
+    manifest = deltacat_storage.get_delta_manifest(delta_locator)
+    delta = Delta.of(delta_locator, None, None, None, manifest)
+    row_count, delta_bytes_pyarrow, delta_manifest_entry_stats = \
+        _calculate_delta_stats(delta, deltacat_storage)
+
+    if StatsType.ROW_COUNT in stat_types_to_collect:
+        delta_stats[StatsType.ROW_COUNT.value] = row_count
+
+    if StatsType.PYARROW_TABLE_BYTES in stat_types_to_collect:
+        delta_stats[StatsType.PYARROW_TABLE_BYTES.value] = delta_bytes_pyarrow
+
+    delta_stats = StatsResult(delta_stats)
+
+    return StatsCompletionInfo.of(delta_locator, delta_stats, delta_manifest_entry_stats)
+
+
+def _get_deltas_from_range(
+        source_partition_locator: PartitionLocator,
+        start_position_inclusive: int,
+        end_position_inclusive: int,
+        deltacat_storage=unimplemented_deltacat_storage) -> List[Delta]:
+
+    namespace, partition_values = source_partition_locator.namespace, source_partition_locator.partition_values
+    table_name, table_version = source_partition_locator.table_name, source_partition_locator.table_version
+    deltas_list_result = deltacat_storage.list_deltas(
+        namespace,
+        table_name,
+        partition_values,
+        table_version,
+        start_position_inclusive,
+        end_position_inclusive,
+        ascending_order=True,
+        include_manifest=False
+    )
+    return deltas_list_result.all_items()
+
+
+def _calculate_delta_stats(delta: Delta,
+                           deltacat_storage=unimplemented_deltacat_storage) -> Tuple[int, int, Dict[int, StatsResult]]:
+    assert delta.manifest is not None, \
+        f"Manifest should not be missing from delta for stats calculation: {delta}"
+    total_rows, total_pyarrow_bytes = 0, 0
+    delta_manifest_entry_stats: Dict[int, StatsResult] = {}
+
+    for file_idx, manifest in enumerate(delta.manifest.entries):
+        entry_pyarrow_table: LocalTable = deltacat_storage.download_delta_manifest_entry(delta, file_idx)
+        assert isinstance(entry_pyarrow_table, pyarrow.Table), \
+            f"Stats collections supported for PyArrow table only"
```

Comment:
To help with debugging, let's also print the unexpected table type:
```suggestion
            f"Stats collection is only supported for PyArrow tables, but received a table of type '{type(entry_pyarrow_table)}' for manifest entry {file_idx} of delta: {delta.delta_locator}."
```
Note that we shouldn't print the whole delta to the logs, since delta manifests may contain megabytes of serialized text data.

Reply:
Yup that makes sense.